### PR TITLE
[WIP] [MNTK] Logging role names; less boilerplatish specs-coding process

### PR DESCRIPTION
**Pull request by Rattenkrieg**
**Not ready for merge**

Solution for #2469 

- [x]  PoC for discovery-time <_node id - node role_> mapping construction
- [x]  Printing/reporting routines adjustment
- [x]  Refactoring Akka.Remote.Tests.MultiNode
- [x]  Refactoring Akka.Cluster.Tests.MultiNode
- [x]  Refactoring Akka.Cluster.Tools.Tests.MultiNode
- [x]  Solution to avoid depending on assemblies being tested in MultiNodeTestRunner
- [x]  Unit-tests for new approach

### CODE REVIEW

**Comment by Rattenkrieg** [comment]
Got rid of remote.tests.multinode references.
[More reflection was employed. ](https://github.com/akkadotnet/akka.net/blob/de510a18de7c1b14f01df5d9f8bdf592866c3373/src/core/Akka.MultiNodeTestRunner/Discovery.cs#L61)
[More command-line arguments were employed.](https://github.com/akkadotnet/akka.net/blob/de510a18de7c1b14f01df5d9f8bdf592866c3373/build.fsx#L322)
Ready for review. Tests on the way.

File: build.fsx [hunk 1]
```diff
@@ -319,6 +319,10 @@ Target "MultiNodeTests" <| fun _ ->
         let args = new StringBuilder()
                 |> append assembly
                 |> append "-Dmultinode.enable-filesink=on"
+                |> append "-Dmultinode.remote-testkit-dll=Akka.Remote.TestKit.dll"
```
Comment (Aaronontheweb) [inline build.fsx]: So what's this for, telling the MNTR where to find the testkit assembly?

File: src/core/Akka.MultiNodeTestRunner/Akka.MultiNodeTestRunner.csproj [hunk 2]
```diff
@@ -73,14 +73,6 @@
       <Project>{964F0EC5-FBE6-47C5-8AE6-145114D5DB8C}</Project>
       <Name>Akka.MultiNodeTestRunner.Shared</Name>
     </ProjectReference>
-    <ProjectReference Include="..\Akka.NodeTestRunner\Akka.NodeTestRunner.csproj">
```
Comment (Aaronontheweb) [inline src/core/Akka.MultiNodeTestRunner/Akka.MultiNodeTestRunner.csproj]: No need for the `NodeRunner.exe` any more?

File: src/core/Akka.MultiNodeTestRunner.Shared/Sinks/MessageSink.cs [hunk 3]
```diff
@@ -88,27 +88,27 @@ public enum MultiNodeTestRunnerMessageType
...
         private const string NodeFailureReasonRegexString =
-            @"\[(\w){4}(?<node>[0-9]{1,2})\]\[(?<status>(FAIL-EXCEPTION))\]{1}\s(?<message>.*)";
+            @"\[(\w){4}(?<node>[0-9]{1,2})(?<role>:\w+)?\]\[(?<status>(FAIL-EXCEPTION))\]{1}\s(?<message>.*)";
         protected static readonly Regex NodeFailureReasonRegex = new Regex(NodeFailureReasonRegexString);
 
         /*
          * Regular expressions - go big or go home. [Aaronontheweb]
          */
-        private const string NodeLogMessageRegexString = @"\[(\w){4}(?<node>[0-9]{1,2})\]\[(?<level>(\w)*)\]\[(?<time>\d{1,4}[- /.]\d{1,4}[- /.]\d{1,4}\s\d{1,2}:\d{1,2}:\d{1,2}(\s(AM|PM)){0,1})\](?<thread>\[(\w|\s)*\])\[(?<logsource>(\[|\w|:|/|\(|\)|\]|\.|-|\$|%|\+|#|\^|@)*)\]\s(?<message>(\w|\s|:|<|\.|\+|>|,|\[|/|-|]|%|\$|\+|\^|@|\(|\))*)";
+        private const string NodeLogMessageRegexString = @"\[(\w){4}(?<node>[0-9]{1,2})(?<role>:\w+)?\]\[(?<level>(\w)*)\]\[(?<time>\d{1,4}[- /.]\d{1,4}[- /.]\d{1,4}\s\d{1,2}:\d{1,2}:\d{1,2}(\s(AM|PM)){0,1})\](?<thread>\[(\w|\s)*\])\[(?<logsource>(\[|\w|:|/|\(|\)|\]|\.|-|\$|%|\+|#|\^|@)*)\]\s(?<message>(\w|\s|:|<|\.|\+|>|,|\[|/|-|]|%|\$|\+|\^|@|\(|\))*)";
```
Comment (Aaronontheweb) [inline src/core/Akka.MultiNodeTestRunner.Shared/Sinks/MessageSink.cs]: We don't really use these `Regex` anymore, now that most node-specific logs are just written directly to their own file

File: src/core/Akka.MultiNodeTestRunner.Shared/Sinks/ConsoleMessageSinkActor.cs [hunk 4]
```diff
@@ -151,17 +151,17 @@ private void WriteSpecMessage(string message)
...
+            Console.WriteLine("[NODE{0}:{1}][{2}]: SPEC PASSED: {3}", nodeIndex, nodeRole, DateTime.UtcNow.ToShortTimeString(), message);
             Console.ResetColor();
         }
 
-        private void WriteSpecFail(int nodeIndex, string message)
+        private void WriteSpecFail(int nodeIndex, string nodeRole, string message)
         {
             Console.ForegroundColor = ConsoleColor.Red;
-            Console.WriteLine("[NODE{0}][{1}]: SPEC FAILED: {2}", nodeIndex, DateTime.UtcNow.ToShortTimeString(), message);
+            Console.WriteLine("[NODE{0}:{1}][{2}]: SPEC FAILED: {3}", nodeIndex, nodeRole, DateTime.UtcNow.ToShortTimeString(), message);
```
Comment (Aaronontheweb) [inline src/core/Akka.MultiNodeTestRunner.Shared/Sinks/ConsoleMessageSinkActor.cs]: Love this change. Having the rolename listed in the log will make things so much easier.

File: src/core/Akka.MultiNodeTestRunner/Akka.MultiNodeTestRunner.csproj (hunk 2 above)
Comment (Rattenkrieg) [inline src/core/Akka.MultiNodeTestRunner/Akka.MultiNodeTestRunner.csproj]: I have deleted this by accident when was cleaning now unnecessary Akka.Remote.TestKit

**Comment by Rattenkrieg** [comment]
Regarding dependency on Akka.Remote.TestKit:
I have implemented discovery routine using only names of types and loading Akka.Remote.TestKit assembly from file passed through commandline. Yesterday everything worked ok, but now when I'm running tests nothing being discovered, and in debug I'm getting weird exceptions. After investigation I have found that Akka.Remote.TestKit is sitting in my Akka.MultiNodeTestRunner/bin folder and interfering with Akka.Remote.TestKit which I'm looking somwhere in <AssemblyUnderTest>/bin. Guess where it came from? 
This is the line from Akka.MultiNodeTestRunner.csproj:
`<Compile Include="..\Akka.Remote.TestKit\CommandLine.cs">`
So we are already depending on Akka.Remote.TestKit but in less obvious way.
I have decided to bring back reflection based on `typeof(MultiNodeConfig)` and sweeping flooded commandline args [here](https://github.com/akkadotnet/akka.net/pull/2502/files/de510a18de7c1b14f01df5d9f8bdf592866c3373#diff-6074983490330a0aac1bd23c958232d7)
Code will be less fragile that way.

File: build.fsx (hunk 1 above)
Comment (Rattenkrieg) [inline build.fsx]: Will remove. For explanation check my comments below.

File: src/core/Akka.MultiNodeTestRunner.Shared/Sinks/MessageSink.cs (hunk 3 above)
Comment (Rattenkrieg) [inline src/core/Akka.MultiNodeTestRunner.Shared/Sinks/MessageSink.cs]: But there is entire _ParsingSpec_ in _Akka.MultiNodeTestRunner.Shared.Tests_ using this regexp.

**Comment by Rattenkrieg** [comment]
Current fails looks transient. Need someone to confirm/re-run.
Everything planned is done, waiting for review.

File: src/core/Akka.MultiNodeTestRunner.Shared/Sinks/MessageSink.cs (hunk 3 above)
Comment (Aaronontheweb) [inline src/core/Akka.MultiNodeTestRunner.Shared/Sinks/MessageSink.cs]: We can remove that spec.

**Comment by Aaronontheweb** [comment]
@Rattenkrieg have a merge conflict with the `ClusterSingletonManagerLeaveSpec`, which was updated in a recent PR.

**Comment by Aaronontheweb** [comment]
OMG, I love this

image

So much easier to figure out WTF is going on now that the role names are included in the per-node and in the aggregate logs.

**Comment by Aaronontheweb** [comment]
@Rattenkrieg so, binary compatibility question: this will require users to upgrade just the `Akka.MultiNodeTestRunner` package they're using, right? Or does the version of the `Akka.Remote.TestKit` they're using also have to be upgraded?

**Comment by Rattenkrieg** [comment]
@Aaronontheweb due to [this](https://github.com/akkadotnet/akka.net/pull/2502/files#diff-c09922238719191ad5e2f7211d0a01e7) change `Akka.Remote.Testkit` is also affected. But this parameter was added mostly for convinience. I can still manipulate with node index on that side though it makes me less confident about order being preserved.

**Comment by Aaronontheweb** [comment]
@Rattenkrieg eh, I don't think that change is the end of the world. Benefits outweigh the costs of upgrading IMHO. 

As part of this PR, could you also update https://github.com/akkadotnet/getakka.net/blob/master/src/docs/testing/multinode-testkit.md ?

**Comment by Rattenkrieg** [comment]
@Aaronontheweb rebased on head of dev, fixed conflicts. Now going to update the docs. As for regex foo you mentioned above is hard to track every caller of `MessageSink` but seems it all ends [here](https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka.MultiNodeTestRunner.Shared/Sinks/SinkCoordinator.cs#L156) and I can't see any `Tell<string>(...)` to that actor. According to that observations it's safe do delete this regex and everything down the road. But I think this is story for separate PR which I can address after finishing this one.

**Comment by Horusiath** [comment]
@Rattenkrieg My question regarding this PR: what scope of changes has affected visualizer? Will it show logs or barrier entries on the timeline? That would be super useful.

**Comment by Rattenkrieg** [comment]
@Horusiath That html file genereated is the product of "Visualizer"? If so then it was affected as everything where I had found usage of nodeid. Check the screenshot below:
default
Don't hesitate to ask for changes if I have missed some logs / or if you think that some logs could be done better.

**Pull request merged**